projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java

```diff
@@ -81,55 +83,76 @@ public class ViewService {
                FILTER NOT EXISTS { ?subject fs:dateDeleted ?anyDateDeleted }
             }
             """, FS.NS));
+    public static final String USER_DOES_NOT_HAVE_PERMISSIONS_TO_READ_FACETS = "User does not have permissions to read facets";
 
     private final Config.Search searchConfig;
+    private final Config.Caches cachesConfig;
     private final ViewsConfig viewsConfig;
     private final Dataset ds;
     private final ViewStoreClientFactory viewStoreClientFactory;
     private final MetadataPermissions metadataPermissions;
-    private final LoadingCache<Boolean, List<FacetDTO>> facetsCache;
-    private final LoadingCache<Boolean, List<ViewDTO>> viewsCache;
+    private LoadingCache<Boolean, List<FacetDTO>> facetsCache;
+    private LoadingCache<Boolean, List<ViewDTO>> viewsCache;
 
-    public ViewService(Config.Search searchConfig,
+    public ViewService(Config config,
                        ViewsConfig viewsConfig,
                        Dataset ds,
                        ViewStoreClientFactory viewStoreClientFactory,
                        MetadataPermissions metadataPermissions) {
-        this.searchConfig = searchConfig;
+        this.searchConfig = config.search;
+        this.cachesConfig = config.caches;
         this.viewsConfig = viewsConfig;
         this.ds = ds;
         this.viewStoreClientFactory = viewStoreClientFactory;
         this.metadataPermissions = metadataPermissions;
-        this.facetsCache = buildFacetCache();
-        this.viewsCache = buildViewsCache();
+        if (cachesConfig.facets.enabled) {
+            this.facetsCache = buildCache(this::fetchFacets, cachesConfig.facets);
+        }
+        if (cachesConfig.views.enabled) {
+            this.viewsCache = buildCache(this::fetchViews, cachesConfig.views);
+        }
         refreshCaches();
     }
 
     public void refreshCaches() {
         log.info("Caches refreshing/warming up has been triggered");
         FilteredDatasetGraph.disableQuadPermissionCheck();
-        facetsCache.refresh(Boolean.TRUE);
-        viewsCache.refresh(Boolean.TRUE);
+        if (cachesConfig.facets.enabled) {
+            facetsCache.refresh(Boolean.TRUE);
+        }
+        if (cachesConfig.views.enabled) {
+            viewsCache.refresh(Boolean.TRUE);
+        }
         FilteredDatasetGraph.enableQuadPermissionCheck();
```

Comment:
not sure if there is any error possible during cache refreshing, but to make sure the `enableQuadPermissionCheck` will be triggered no matter what, maybe it might be better to wrap the refreshing in try/catch and do `enableQuadPermissionCheck` in `finally`?

Reply:
very good catch, super makes sense, added